Add regex for time format to `resource_compute_resource_policies`

karolgorc opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->
closes [issue #8460](https://github.com/hashicorp/terraform-provider-google/issues/8460)

This is probably a breaking change because some `terraform apply` commands won't go through if they have a wrong format. The regex will throw an error. But it also ensures that it's the correct format for the API.

Will write a test case for this into the PR if it's an approved solution.

<!--
Please self-review your PR against the review checklist before creating it: https://googlecloudplatform.github.io/magic-modules/contribute/review-pr/

Completing the checklist will help speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.

If your PR is still work in progress, please create it in draft mode
-->

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none

Unless you choose release-note:none, please add a release note.

See https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/ for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:breaking-change
compute: `resource_policies` now have a regex on their time format that will allow for correct calls to the API
```

modular-magician commented:
Hi there, I'm the Modular magician. I've detected the following information about your changes:

## Diff report

Your PR generated some diffs in downstreams - here they are.

`google` provider: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-11486-old..auto-pr-11486) ( 1 file changed, 2 insertions(+), 2 deletions(-))
`google-beta` provider: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-11486-old..auto-pr-11486) ( 1 file changed, 2 insertions(+), 2 deletions(-))

modular-magician commented:
#### Tests analytics
Total tests: 982
Passed tests: 908
Skipped tests: 72
Affected tests: 2

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>

#### Action taken
<details>
<summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeDisk_storagePoolSpecified</li>
<li>TestAccComputeInstance_bootDisk_storagePoolSpecified</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{red}{\textsf{Tests failed during RECORDING mode:}}$
`TestAccComputeDisk_storagePoolSpecified`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/build-log/recording_build/TestAccComputeDisk_storagePoolSpecified_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/recording/TestAccComputeDisk_storagePoolSpecified.log)]
`TestAccComputeInstance_bootDisk_storagePoolSpecified`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/build-log/recording_build/TestAccComputeInstance_bootDisk_storagePoolSpecified_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/recording/TestAccComputeInstance_bootDisk_storagePoolSpecified.log)]


$\textcolor{red}{\textsf{Errors occurred during RECORDING mode. Please fix them to complete your PR.}}$


View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/d4e9c0a0-612b-4b8d-b4ac-8e0dba5a95d7/recording) for each test

SarahFrench commented on mmv1/products/compute/ResourcePolicy.yaml:
```diff
@@ -142,7 +142,7 @@ properties:
                 required: true
                 validation: !ruby/object:Provider::Terraform::Validation
-                  function: 'verify.ValidateHourlyOnly'
+                  regex: '^([01][0-9]|2[0-3]):00$'
```
Note for posterity:

I tested the regex and can see that these cases are all identified as invalid

- 0:00
- 0:01
- 1:00
- 10:10
- 24:00

and these are identified as valid:

- 01:00
- 10:00
- 20:00
- 23:00

SarahFrench reviewed (commented):
Hi, thanks for your PR!

My understanding of the problem is that currently if a user was to use "1:00" in their configuration Terraform will identify a permadiff because the API will return "01:00" and there's a mismatch as a result.

The new regex for validating input looks correct to me, but you're also correct that this would be a breaking change for some users - e.g. some might have used [ignore_changes](https://developer.hashicorp.com/terraform/language/meta-arguments/lifecycle#ignore_changes) on this field to navigate the permadiff.

Another approach is to add [diff suppression to the field](https://googlecloudplatform.github.io/magic-modules/develop/field-reference/#diff_suppress_func) so that the provider regards "1:00" and "01:00" (for example) as equivalent and doesn't raise a diff in this situation. Would you consider this alternative instead, as it's non-breaking?

karolgorc commented:
ok i wasn't aware that the Google API actually translated the given value from H:MM to HH:MM. If that happens then the suppress diff seems like a better solution

- Added a HourlyFormatSuppressDiff() to `start_time` under `hourly_schedule` and `daily_schedule`
- Added unit test for it

SarahFrench reviewed (commented):
>@SarahFrench This PR has been waiting for review for 3 weekdays. Please take a look!

I'm OOO and will review this on Tuesday

SarahFrench reviewed (commented):
This looks good to me, thanks for the testing of the diff suppression function! Before merging I'd like there to be an acceptance test that contains a `start_time` value in H:mm format and shows that there aren't errors.

An approach for this could be to update [this example file](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/examples/resource_policy_basic.tf.erb), which is referenced here in [ResourcePolicy.yaml ](https://github.com/GoogleCloudPlatform/magic-modules/blob/a6206ac9911804d06cf156ca028f33e5159a95bc/mmv1/products/compute/ResourcePolicy.yaml#L45-L49), so that the `start_time` value can be passed in from the YAML as a variable (similar to what you see for `name`). After doing this, you could add another entry to the `examples` list in ResourcePolicy.yaml but set the value for `start_time` in the H:mm format. You'd also want to include `skip_docs: true` in that block, so that the example is only used to generate a test and doesn't accidentally promote using H:mm format format in the official provider documentation

karolgorc commented:
Using the same `example` file generates 2 test functions that have the same name. If there is a way to generate 2 unique testing functions using the same `example` file please LMK! But for now i just created another file that stores the new config with `4:00` instead of `04:00`.

```
--- PASS: TestAccComputeResourcePolicy_resourcePolicyHourlyFormatExample (12.39s)
PASS
ok      github.com/hashicorp/terraform-provider-google/google/services/compute    12.417s
```

SarahFrench commented:
Ah, of course name collision would happen 🤦 . In that case I'm afraid that the route forward is to create a new example file with a different name that has the start_time value set in H:mm. Because of this having the time set via a variable wouldn't be necessary. Sorry for leading you down an incorrect path!

karolgorc commented:
>In that case I'm afraid that the route forward is to create a new example file with a different name that has the start_time value set in H:mm

Done [here](https://github.com/GoogleCloudPlatform/magic-modules/pull/11486/commits/ec4da7dde87d6fa9e498caa29f0742524d778312)

modular-magician commented:
#### Tests analytics
Total tests: 990
Passed tests: 911
Skipped tests: 77
Affected tests: 2

<details>
<summary>Click here to see the affected service packages</summary>
<blockquote>

<ul>
<li>compute</li>

</ul>

</blockquote>
</details>
#### Non-exercised tests

Tests were added that are skipped in VCR:
- TestHourlyFormatSuppressDiff
#### Action taken
<details>
<summary>Found 2 affected test(s) by replaying old test recordings. Starting RECORDING based on the most recent commit. Click here to see the affected tests
</summary>
<blockquote>
<ul>
<li>TestAccComputeInstanceTemplate_withNamePrefix</li>
<li>TestAccComputeResourcePolicy_resourcePolicyHourlyFormatExample</li>

</ul>
</blockquote>
</details>

[Get to know how VCR tests work](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/contributing/#general-contributing-steps)

modular-magician commented:
$\textcolor{green}{\textsf{Tests passed during RECORDING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]
`TestAccComputeResourcePolicy_resourcePolicyHourlyFormatExample`[[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/recording/TestAccComputeResourcePolicy_resourcePolicyHourlyFormatExample.log)]
$\textcolor{red}{\textsf{Tests failed when rerunning REPLAYING mode:}}$
`TestAccComputeInstanceTemplate_withNamePrefix`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/build-log/replaying_build_after_recording/TestAccComputeInstanceTemplate_withNamePrefix_replaying_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/replaying_after_recording/TestAccComputeInstanceTemplate_withNamePrefix.log)]


Tests failed due to non-determinism or randomness when the VCR replayed the response after the HTTP request was made.

Please fix these to complete your PR. If you believe these test failures to be incorrect or unrelated to your change, or if you have any questions, please raise the concern with your reviewer.


---




$\textcolor{green}{\textsf{All tests passed!}}$

View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-11486/artifacts/9bfb5274-ff82-4259-b6b7-034971df54d1/recording) for each test

SarahFrench reviewed (approved):
Thanks, and sorry again! The new generated test passes so I'll merge. The `TestAccComputeInstanceTemplate_withNamePrefix` failure is unrelated to this PR.
